Get WITH_PARQUET=ON working in CI

pjones opened the pull request:
## Description

This PR enables `WITH_PARQUET` on all platforms and resolves issues that prevented the CI builds from finishing successfully.

The primary issue was related to the `CMAKE_PREFIX_PATH` environment variable getting mangled during its travels from the CI YAML file, through CTest to CMake and finally to the `CMakeCache.txt` file.  Once I fixed that problem, several past workarounds broke, which needed to be fixed as well.

Note: I tried just removing `CMAKE_PREFIX_PATH` from showing up in the `CMakeCache.txt` file but that didn't work.  Some weird interaction between CTest and CMake prevented that variable from having a semicolon-separated list of paths.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Parquet support enabled in more build/test contexts.

* **Chores**
  * Streamlined CI/setup across Windows, macOS, and Linux; improved tool installation and PATH handling.
  * Added environment capture before test runs to record and propagate build-related variables.
  * Broadened propagation of build-prefix outputs to improve library discovery during builds.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

pjones committed:
cmake: Don't mangle cached environment variables

When using CMake/CTest to write out a `CMakeCache.txt` file,
environment variables injected into a string variable, which is then
written to a file via CMake, results in the value of those variables
being mangled.

For example, CMake list variables such as `CMAKE_PREFIX_PATH` have
their semicolon separators removed, resulting in an invalid list and
errors downstream about missing libraries.

I'm not sure it's wise to do this in the first place, but I kept this
cache initialization system and fixed the issue by moving it out into
a shell script.

pjones committed:
parquet: Re-enable parquet support and refactor the conda/CI interaction

Set `WITH_PARQUET` default value back to `ON`.

This is mostly a refactor to make the Conda steps more universal.
That is, for other operating systems and possibly other libraries that
we might need in the future.

pjones committed:
windows/qt: Don't expose library path twice

Primary change: Only export the Qt path once.  This didn't work in the
past but was hidden by the "mangled CMake variables" bug.  This fixed
that bug.

But, I couldn't help but clean up this section so that items are
properly grouped together (e.g., choco installs).  Like all other
sections, `echo` statements that affect the following steps are put at
the end of the run block.

pjones committed:
cmake: CMake uses semicolons as list separators

pjones commented on .github/workflows/openms_ci_matrix_full.yml:
```diff
@@ -434,7 +455,7 @@ jobs:
           #  Problem: not all compilers are always in the PATH.
           CMAKE_CXX_COMPILER: ${{ steps.set-vars.outputs.cxx_compiler }}
-          CMAKE_PREFIX_PATH: ${{ steps.tools-prefix.outputs.cmake_prefix }}
+          CMAKE_PREFIX_PATH: "${{ steps.tools-prefix.outputs.cmake_prefix }};${{ steps.conda-setup.outputs.cmake_prefix }}"
```
CMake can deal with this.  No need to introduce the extra complexity to fix a non-problem.

jpfeuffer commented:
Looks good. Just some elaboration on why it did not work before.
I removed CMAKE_PREFIX_PATH from the cache such that CMake solely reads it from the environment (I am still not sure what CMake does if it now sees the CMAKE_PREFIX_PATH in the environment AND in the cache.
The problem is, that CMake expects them in different list formats depending on how you specify them. In the environment it assumes it is is a PATH-like variable and separated by the OS path separator. As a cache variable or when passed to CMake via -D, it expects it as a CMake lists (always separated by semicolon).

In theory I don't care how it is done if it is correct but keep in mind that in theory you now have CMAKE_PREFIX_PATH in both the shell environment as well as in the CMakeCache. It looks like the cache rightfully gets preference so fine with me.

jpfeuffer commented:
Note: There might still be problems during creating installers, right? Is that next?

pjones commented:
> Note: There might still be problems during creating installers, right? Is that next?

Nope, this fixes that problem.

pjones commented:
> Looks good. Just some elaboration on why it did not work before. I removed CMAKE_PREFIX_PATH from the cache such that CMake solely reads it from the environment (I am still not sure what CMake does if it now sees the CMAKE_PREFIX_PATH in the environment AND in the cache. The problem is, that CMake expects them in different list formats depending on how you specify them. In the environment it assumes it is is a PATH-like variable and separated by the OS path separator. As a cache variable or when passed to CMake via -D, it expects it as a CMake lists (always separated by semicolon).
> 
> In theory I don't care how it is done if it is correct but keep in mind that in theory you now have CMAKE_PREFIX_PATH in both the shell environment as well as in the CMakeCache. It looks like the cache rightfully gets preference so fine with me.

@jpfeuffer 

I tried to get CMake to pick up `CMAKE_PREFIX_PATH` from the environment and it was still messed up some how.  I *think* it may have something to do with CTest shelling out to CMake, but I'm not sure.  Perhaps it has something to do with the path separator, but that should be a semicolon too on Windows.  

I also tried having CTest pass `-DCMAKE_PREFIX_PATH=$ENV{CMAKE_PREFIX_PATH}` but that also failed.  The semicolons were stripped out and what was left was an invalid path.  I don't really know what the problem is but once I moved it out of CTest/CMake everything started working as expected.
